### ISSUE TITLE
Custom Output Provider and retries over transient failures

### ISSUE THREAD

**Issue by SixFootRhino** [Azure/diagnostics-eventflow#237]
Perhaps I'm overlooking something, but the documentation isn't clear as to what happens over time if our output provider is failing to transmit, and if we can rely on EF engine at all to making solving this problem easier when trying to guarantee delivery of events.

In short, will we ever see the same events multiple times if we throw an exception while processing them the first or subsequent go-around?

For example, let's say we are forwarding to an external Web API. Or even to a Service Fabric reliable service, and it is temporarily unavailable. Do we have to queue these up ourselves and build our own async retry mechanism to guarantee delivery?

Or can we do something like throw an exception and EF will re-present the items to us in the next call with an increased batch size? For example:

- EF calls the output provider with N items
- output provider throws an exception after a timeout to external system
- EF as accumulated M additional events in the meantime
- on next call EF sends a batch with N + M items (up to batch size limit)

That would seem very logical solution to a common problem. But it does potentially bring up other issues of what to do about ever increasing memory consumption when providers continually fail. So I believe this would either need to be a provider capability that you could query, or, you could require that as specific exception be thrown to drive this.

Either way, In our cases, we're fine with this since the downstream systems are reliable, and in fact we *want* these EF queues to grow and grow and not drop events due to transient issues.

Thanks in advance for answering these!

**Comment by karolz-ms** [comment]
Hey @SixFootRhino sorry for late reply. 

All good ideas but... the current EF engine is not doing any retries if the output fails. So it all depends on how the output is implemented. The core engine will try just once. If the transmission fails, it will tell the health reporter (emit a warning) and move on.

Among standard outputs only Application Insights output has any retry capability, and that is only when [server telemetry channel](https://github.com/Microsoft/ApplicationInsights-dotnet/tree/develop/src/ServerTelemetryChannel) is used. It is a fairly tricky piece of code that involves filesystem-based caching. I do not think it is likely that something like this will appear in core EF in foreseeable future.

**Comment by SixFootRhino** [comment]
Hey thanks for getting back to me.

I figured as much, and already had a plan for how to address this. We were just hoping to avoid the work. ;-)

In our case, we have a store-and-forward batching component that we wrote for an auditing stateful service. It is backed by a reliable queue that, today, writes-through to a postgres database. But it's designed such that it could "forward" to any system.

We can re-use that component in our custom OutputProvider so that no events are ever "lost". But I do have concerns on impact of performance with putting all of those events into a reliable collection. 

However, that's a problem for future me. ;-)
